pkg/k8s/resource/event.go

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package resource
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	k8sRuntime "k8s.io/apimachinery/pkg/runtime"
+)
+
+// Event emitted from resource. One of SyncEvent, UpdateEvent or DeleteEvent.
+type Event[T k8sRuntime.Object] interface {
+	// Handle the event by invoking the right handler.
+	// On error the event is requeued by key for later processing.
+	//
+	// If you use Handle(), then Done() should not be called.
+	// If you need to process the events in parallel/asynchronously,
+	// then do a type-switch on Event[T] and call Done() after the
+	// event has been processed.
+	Handle(
+		onSync func(Store[T]) error,
+		onUpdate func(Key, T) error,
+		onDelete func(Key, T) error,
+	)
+
+	// Done marks the event as processed.  If err is non-nil, the
+	// key of the object is requeued and the processing retried at
+	// a later time with a potentially new version of the object.
+	//
+	// If you choose not to use Handle(), then this must always be called after the
+	// event has been processed.
+	Done(err error)
+}
+
+type baseEvent struct {
+	done func(error)
+}
+
+func (b *baseEvent) Done(err error) {
+	b.done(err)
+}
+
+// SyncEvent is emitted when the store has completed the initial synchronization
```

Comment:
What does this mean, exactly? Does it indicate simply that the store is loaded? Does it mean that every pre-existing item has been Observed at least once?

Reply:
This means that the set of updated objects preceding the sync event constitute a complete snapshot and that the store can now be used (e.g. ListKeys() will not return a partially synced view). 

In practical terms we need the sync event for garbage collection purposes. For example when the agent starts we would synchronize state to BPF maps and then run a garbage collection to see if there are entries in the BPF maps that no longer have a corresponding k8s object.

I'll try to improve the documentation!